File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
class Solution {
public:
    int maxi=INT_MIN;
    int dfs(TreeNode* root){
        if(!root)return 0;
        int left=max(dfs(root->left),0);
        int right=max(dfs(root->right),0);
        maxi=max(maxi,root->val+left+right);
        return max(root->val+left, root->val+right);  //we are returning single_Path_Sum and we already covered complete path sum(i.e. root->val+left+right ) in line number 8;
        
    }
    int maxPathSum(TreeNode* root) {
        if(!root)return 0;
        dfs(root);
        return maxi;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
class Solution(object):
    current_max = float('-inf')
    def maxPathSum(self, root):
        self.maxPathSumHelper(root)
        return self.current_max

    def maxPathSumHelper(self, root):
        """Helper method"""
        if root is None:
            return root
        left = self.maxPathSumHelper(root.left)
        right = self.maxPathSumHelper(root.right)
        left = 0 if left is None else (left if left > 0 else 0)
        right = 0 if right is None else (right if right > 0 else 0)
        self.current_max = max(left+right+root.val, self.current_max)
        return max(left, right) + root.val